src/mod/endpoints/mod_gsmopen/gsmlib/gsmlib-1.10-patched-13ubuntu/doc/gsmpb.man
.\" -*- eval: (nroff-mode) -*-
.de TQ
.br
.ns
.TP \\$1
..
.\" Like TP, but if specified indent is more than half
.\" the current line-length - indent, use the default indent.
.de Tp
.ie \\n(.$=0:((0\\$1)*2u>(\\n(.lu-\\n(.iu)) .TP
.el .TP "\\$1"
..
.TH GSMPB 8 "##DATE##" "gsmpb v##VERSION##"
.SH NAME
gsmpb \- GSM mobile phone phonebook manipulation program
.SH SYNOPSIS
.B gsmpb
[ \fB\-b\fP \fIbaudrate\fP ]
[ \fB\-\-baudrate\fP \fIbaudrate\fP ]
[ \fB\-c\fP ]
[ \fB\-\-copy\fP ]
[ \fB\-d\fP \fIdestination device or file\fP ]
[ \fB\-\-destination\fP \fIdestination device or file\fP ]
[ \fB\-h\fP ]
[ \fB\-\-help\fP ]
[ \fB\-i\fP ]
[ \fB\-\-index\fP ]
[ \fB\-I\fP \fIinit string\fP ]
[ \fB\-\-init\fP \fIinit string\fP ]
[ \fB\-p\fP \fIphonebook name\fP ]
[ \fB\-\-phonebook\fP \fIphonebook name\fP ]
[ \fB\-s\fP \fIsource device or file\fP ]
[ \fB\-\-source\fP \fIsource device or file\fP ]
[ \fB\-t\fP \fIcharacter set\fP ]
[ \fB\-\-charset\fP \fIcharacter set\fP ]
[ \fB\-v\fP ]
[ \fB\-\-version\fP ]
[ \fB\-V\fP ]
[ \fB\-\-verbose\fP ]
[ \fB\-X\fP ]
[ \fB\-\-xonxoff\fP ]
[ \fB\-y\fP ]
[ \fB\-\-synchronize\fP ]
.PP
.SH DESCRIPTION
\fIgsmpb\fP can store or retrieve phonebook entries residing in a GSM
mobile phone's phonebook to or from a file. A synchronization mode is
also available.
.PP
\fIgsmpb\fP reads entries from the source which can be a mobile phone
(if a serial device file is given) or a file (if a file name is
given). The source is never modified. \fIgsmpb\fP writes phonebook
entries to a destination file or device.  Depending on the mode the
source is copied to the destination file, thus overwriting the
destination, or the destination is synchronized with regard to the
source which is the default (details see below).
.PP
If "\-" is given as the parameter for the \fB\-\-source\fP or
\fB\-\-destination\fP options, the phonebook is read from standard input 
and/or written to standard output, respectively.
.PP
Phonebook entries names are encoded using the GSM default alphabet in
the mobile phone, whereas they are stored using the Latin\-1 encoding
in phonebook files. When reading phonebook entries from a mobile phone 
entry names are converted from the GSM default to Latin\-1. Characters
that can not be converted to Latin\-1 are encoded as character code
172 (Latin\-1 boolean "not"). When writing file-based phonebook entries 
to a mobile phone a conversion to the GSM default alphabet takes
place. Characters that can not be converted are encoded as GSM delta
(code 16). If the default character set has been changed using the
\fB\-\-charset\fP option no conversion takes place.
.PP
Error messages are printed to the standard error output. If the program
terminates on error the error code 1 is returned.
.PP
.SH OPTIONS
.TP .7i
\fB\-b\fP \fIbaudrate\fP, \fB\-\-baudrate\fP \fIbaudrate\fP
The baud rate to use. The default baudrate is 38400.
.TP .7i
\fB\-c\fP, \fB\-\-copy\fP
This causes the contents of the source to be copied to the
destination. After this operation the destination has exactly the same 
contents as the source.
.TP .7i
\fB\-d\fP \fIdestination\fP, \fB\-\-destination\fP \fIdestination\fP
The destination device or file.
.TP .7i
\fB\-h\fP, \fB\-\-help\fP
Prints an option summary.
.TP .7i
\fB\-I\fP \fIinit string\fP, \fB\-\-init\fP \fIinit string\fP
Initialization string to send to the TA (default: "E0"). Note that the
sequence "ATZ" is sent first.
.TP .7i
\fB\-i\fP, \fB\-\-index\fP
If the index position is given, \fIgsmpb\fP preserves the assignment
of entries to memory slots in the mobile phone's phonebook. This can
be used to backup phonebook entries with their position into a
phonebook file or to change the position of entries by editing a
phonebook file and writing them back to the mobile phone.
If this option is given the phonebook file used as the source
must contain indices for every entry. Additionally, these indices must 
be unique, ie. it is not allowed to assign one entry twice to a
specific position in the mobile phone's phonebook.
.TP .7i
\fB\-p\fP \fIphonebook\fP, \fB\-\-phonebook\fP \fIphonebook\fP
The name of the phonebook to read from or write to. This is only used
for device sources and destinations. Commonly available phonebooks
are:
.TP .3i
.po +0.7i
.ll 5.8i
\fIFD\fP
SIM fixdialling\-phonebook
.TP .3i
\fILD\fP
SIM last\-dialling\-phonebook
.TP .3i
\fIME\fP
ME phonebook
.TP .3i
\fIMT\fP
combined ME and SIM phonebook
.TP .3i
\fISM\fP
SIM phonebook
.TP .3i
\fITA\fP
TA phonebook
.TP .7i
.po -0.7i
.ll 6.5i
\fB\-s\fP \fIsource\fP, \fB\-\-source\fP \fIsource\fP
The source device or file.
.TP
\fB\-t\fP \fIcharacter set\fP, \fB\-\-charset\fP \fIcharacter set\fP
Set the character set to use for phonebook operations (default is the
GSM default alphabet).
.TP
\fB\-v\fP, \fB\-\-version\fP
Prints the program version.
.TP .7i
\fB\-V\fP, \fB\-\-verbose\fP
Prints out a detailed progress report.
.TP .7i
\fB\-X\fP, \fB\-\-xonxoff\fP
Uses software handshaking (XON/XOFF) for accessing the device.
.TP .7i
\fB\-y\fP, \fB\-\-synchronize\fP
This causes the contents of the source to be synchronized with the
destination (default). Synchronization in this context means:
.TP .2i
.po +0.7i
.ll 5.8i
\-
If the source contains an entry with a name that does not exist in the 
destination this entry is added to the destination.
.TP .2i
\-
If the source contains an entry with a name that can also be found in
the destination, the entry in the destination is overwritten (ie. the
telephone number is updated). Exception: More then one entry with the
name exists in the destination. In this case the new entry ist just added.
.TP .2i
\-
Entries in the destination that do not exist in the source are
deleted.
.PP
Note that synchronization has the following properties that differ
from copying: This algorithm does not change the location of unchanged
entries in the destination phonebook. The synchronization function
is not case-sensitive when comparing names.
.PP
.po -0.7i
.ll 6.5i
.SH PHONEBOOK FILE FORMAT 
Phonebook entries are stored in phonebook files that are meant to be
human-readable and -editable. There is one phonebook entry per line,
and each line has the format:
.PP
.nf
index|text|phone number
.fi
.PP
The fields have the following meanings:
.TP .7i
\fIindex\fP
The index of the entry which must be a positive number. The index may
also be empty. Indices can be used in conjunction with the
\fB\-\-index\fP option to store the entry into a specific position in
the mobile phone.
.TP .7i
\fItext\fP
Descriptive text for the entry. The text may contain the special
characters '\\', '|', carriage return (ASCII code 13), or line feed
(ASCII code 10). These must be written "\\\\", "\\|", "\\r", "\\n",
respectively. The text should only contain characters that can be
encoded using the GSM default alphabet (see comments above).
.TP .7i
\fIphone number\fP
Phone numbers can only contains the digits 0\-9 and the '+' sign. A '+' 
sign denotes an international number.
.PP
.SH EXAMPLES
The following invocation of \fIgsmpb\fP synchronizes the mobile phone's
SIM phonebook with the file $HOME/.phonebook:
.PP
.nf
gsmpb \-\-synchronize \-b 19200 \-d /dev/mobilephone \\
      \-s $HOME/.phonebook \-p "SM"
.fi
.PP
.SH AUTHOR
Peter Hofmann <software@pxh.de>
.PP
.SH BUGS
Report bugs to software@pxh.de.  Include a complete, self-contained
example that will allow the bug to be reproduced, and say which
version of \fIgsmpb\fP you are using.
.PP
.SH COPYRIGHT
Copyright \(co 1999 Peter Hofmann
.LP
\fIgsmpb\fP is free software; you can redistribute it and/or modify it under
the terms of the GNU Library General Public License as published by the Free
Software Foundation; either version 2, or (at your option) any later
version.
.LP
\fIgsmpb\fP is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public License
for more details.
.LP
You should have received a copy of the GNU Library General Public License along
with \fIgsmpb\fP; see the file COPYING.  If not, write to the Free Software
Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.PP
.SH "SEE ALSO"
.BR gsminfo(7),
.BR gsmctl(1),
.BR gsmsendsms(1),
.BR gsmsmsd(8),
.BR gsmsmsstore(1).



ChangeLog
 * libZRTP SDK library, implements the ZRTP secure VoIP protocol.
 * Copyright (c) 2006-2009 Philip R. Zimmermann.  All rights reserved.
 * Contact: http://philzimmermann.com
 * For licensing and other legal details, see the file zrtp_legal.c.

!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
SINCE LIBZRTP v0.80 CHANGELOG IS A PART OF HTML DOCUMENTATION. 
https://developers.zfoneproject.com/libzrtp/wiki/LibzrtpChangeLog
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!


libzrtp 0.7.1											18.11.2008
--------------------------------------------------------------------------------
1.	Fixed bug with hardcoded AES128 cipher for generating SRTP keys. In this 
	version is selected according to ZRTP discovery.
	
2.	Added initialization/deinitalization functions to zrtp helper functions.

libzrtp 0.7.0											04.11.2008
--------------------------------------------------------------------------------
1.	Changes in libzrtp sources tree.

2.	Improvements in libzrtp initialization routine:
	- all global zrtp options were combined in zrtp_config_t structure;
	- zrtp_init() allocates memory for zrtp global context;
	- zrtp_config_defaults() 

3.	Improvements for Passive/Active mode support.
	a) A passive endpoint never sends a commit message, period. Also, it 
	declares itself as a passive endpoint by setting the P flag it its own
	Hello message;
	b) A active endpoint does not send a commit to a passive endpoint, which it 
	recognizes by detecting the P flag;
	c) A passive phone, if acting as a SIP initiator (meaning it initiated the 
	call), rejects all commit packets from everyone;
	d) A passive phone rejects all commit messages from a PBX, which is easily 
	recognized by the M flag.
	Passive mode support is built into the library logic and will be used
	automatically if the developer specifies signaling role by setting 
	is_initiator flag in zrtp_init_session_ctx().

4.	Improvements in ZRTP feedback interface and system-dependent functions.
    There are two types of interface functions in libzrtp: system dependent API
	and realization of helper functions and events. System dependent API in
	defined in zrtp_iface_system.h as set of extern functions. System
	functions are already implemented for several basic platforms in
	zrtp_iface.sys.c libzrtp feedback and helper functions were re-factored
	and implemented as set of callbacks. If the developer doesn't want to
	handle one or another event it may just leave necessary pointer empty.
	See zrtp_init() and zrtp_callback_t, zrtp_iface.h for more details.

5. 	ZRTP configuration approach was improved: zrtp_config_xxx.h contains
	adjustments for necessary target platform. libzrtp contains default
	configs for Linux, OS X, Windows, Window CE and Symbian platforms.
	All ZRTP protocol and behavior related adjustments are collected in
	zrtp_config_user.h. Edit this file to configure libzrtp for your
	needs.
	
6.	Implemented new functions in protocol according to the Internet Draft v 10.

7.	Improved realization of built-in libzrtp scheduler. Fixed bug with crashing
	on performing delay call when zrtp session have been already deleted.
	
8. 	Logging function was improved. Use ZRTP_LOG macro to print log messages. See
	zrtp_log.h for more information.


libzrtp 0.6.8											03.09.2008
--------------------------------------------------------------------------------
ZRTP
1.	Fixed bug with incorrect maximum value for T1 retry interval timer which 
    increased delay between LOOKING_FOR_ZRTP and NO_ZRTP_SUPPORT states. 
    Internal fix, no API changes required;
2.	Implemented version negotiation according to the latest specification.
	libzrtp v 0.6.8 supports ZRTP v0.90 only. No changes required in 
	applications that use the SDK.
3. 	Compilation flag WITH_ZFONE was removed. The developer, who wants to use
	built-in ZRTP cache, has to set name of the ZRTP cache explicitly,
	implementing  zrtp_get_cache_path() function.
4. 	New libzrtp licensing scheme was implemented. It allows the licensing policy 
    to be changed at run time.  See zrtp_license_mode_t doc. for more details. 
    Affected API - zrtp_init().
5. 	ZRTP Protocol version was changed to 0.90 according to ZRTP Internet Draft.
6. 	Some changes in Linux config files: surplus configuration flags were removed
	from ./cfg.XXX templates. 
7.	Added experimental ZRTP messages retries scheduler for slow channels. As 
	ex ample for GSM CSD channel with average bandwidth 6Kb/s. To use this
	option build library with BUILD_FOR_CSD flag.


libzrtp 0.6.6											27.06.2008
--------------------------------------------------------------------------------
ZRTP
1. Small bug was fixed in S0 calculation: when RS1 is corrupted the library uses
   RS2 instead;
2. Some changes in Makefile and building process: unused header were eliminated 	
   from the installation process.
3. -DBUILD_WITH_ZRTP_MUTEXES was replaced with --enable_mutexes option passed to 
   ./configure script. This change allows not to specify any libzrtp compilation 
   flags during user application compilation. --enable-mutexes adds    
   BUILD_ZRTP_MUTEXES definition to the ./config/zrtp_unix_config.h so if you 
   build libzrtp on other platforms - define this flag manually (windows 
   configuration file already includes this option).
4. Clean-up in .h and .c comments was made.


libzrtp 0.6.5											04.06.2008
--------------------------------------------------------------------------------
ZRTP
1. New names for: other_secret - pbxs; srtps - auxs. In bits and secrets storages;
2. RS2 secret was eliminated form DH s0 calculation;
3. Protocol version number was increased to 0.85


libzrtp 0.6.4											19.05.2008
--------------------------------------------------------------------------------
ZRTP
1. According to the new version of the Internet Draft Signaling shared secret was
   removed from the protocol and from the sources. It was not used by interface

src/mod/endpoints/mod_gsmopen/gsmlib/gsmlib-1.10-patched-13ubuntu/apps/gsmpb.cc
// *************************************************************************
// * GSM TA/ME library
// *
// * File:    gsmpb.cc
// *
// * Purpose: phonebook management program
// *
// * Author:  Peter Hofmann (software@pxh.de)
// *
// * Created: 24.6.1999
// *************************************************************************

#ifdef HAVE_CONFIG_H
#include <gsm_config.h>
#endif
#include <gsmlib/gsm_nls.h>
#include <string>
#ifdef WIN32
#include <gsmlib/gsm_win32_serial.h>
#else
#include <gsmlib/gsm_unix_serial.h>
#include <unistd.h>
#endif
#if defined(HAVE_GETOPT_LONG) || defined(WIN32)
#include <getopt.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <gsmlib/gsm_me_ta.h>
#include <gsmlib/gsm_util.h>
#include <gsmlib/gsm_sorted_phonebook.h>
#include <iostream>

using namespace std;
using namespace gsmlib;

#ifdef HAVE_GETOPT_LONG
static struct option longOpts[] =
{
  {"xonxoff", no_argument, (int*)NULL, 'X'},
  {"phonebook", required_argument, (int*)NULL, 'p'},
  {"init", required_argument, (int*)NULL, 'I'},
  {"destination", required_argument, (int*)NULL, 'd'},
  {"source", required_argument, (int*)NULL, 's'},
  {"destination-backend", required_argument, (int*)NULL, 'D'},
  {"source-backend", required_argument, (int*)NULL, 'S'},
  {"baudrate", required_argument, (int*)NULL, 'b'},
  {"charset", required_argument, (int*)NULL, 't'},
  {"copy", no_argument, (int*)NULL, 'c'},
  {"synchronize", no_argument, (int*)NULL, 'y'},
  {"help", no_argument, (int*)NULL, 'h'},
  {"version", no_argument, (int*)NULL, 'v'},
  {"verbose", no_argument, (int*)NULL, 'V'},
  {"indexed", no_argument, (int*)NULL, 'i'},
  {(char*)NULL, 0, (int*)NULL, 0}
};
#else
#define getopt_long(argc, argv, options, longopts, indexptr) \
  getopt(argc, argv, options)
#endif

// insert those entries from sourcePhonebook into destPhonebook
// that are not already present in destPhonebook

void insertNotPresent(SortedPhonebookRef sourcePhonebook,
                      SortedPhonebookRef destPhonebook,
                      bool indexed, bool verbose)
{
  for (SortedPhonebookBase::iterator i = sourcePhonebook->begin();
       i != sourcePhonebook->end(); ++i)
  {
    pair<SortedPhonebookBase::iterator, SortedPhonebookBase::iterator> range;
    if (indexed)
    {
      int index = i->index();
      range = destPhonebook->equal_range(index);
    }
    else
    {
      string text = i->text();
      range = destPhonebook->equal_range(text);
    }

    // do nothing if the entry is already present in the destination
    bool alreadyPresent = false;
    for (SortedPhonebookBase::iterator j = range.first;
         j != range.second; ++j)
    {
      i->setUseIndex(indexed);
      if (i->telephone() == j->telephone())
      {
        alreadyPresent = true;
        break;
      }
    }
    // ... else insert it
    if (! alreadyPresent)
    {
      if (verbose)
      {
        cout << stringPrintf(_("inserting '%s' tel# %s"),
                             i->text().c_str(), i->telephone().c_str());
        if (indexed)
          cout << stringPrintf(_(" (index #%d)"), i->index());
        cout << endl;
      }
      i->setUseIndex(indexed);
      destPhonebook->insert(*i); // insert
    }
  }
}

// update those entries in destPhonebook, that
// - have the same name as one entry in destPhonebook
// - but have a different telephone number
// this is only done if the name in question is unique in the destPhonebook
// the case of several entries having the same in the sourcePhonebook
// is handled - only the first is considered for updating

void updateEntries(SortedPhonebookRef sourcePhonebook,
                   SortedPhonebookRef destPhonebook,
                   bool verbose)
{
  bool firstLoop = true;
  string lastText;

  for (SortedPhonebookBase::iterator i = sourcePhonebook->begin();
       i != sourcePhonebook->end(); ++i)
  {
    string text = i->text();
    if (! firstLoop && text != lastText)
    {
      pair<SortedPhonebookBase::iterator,
        SortedPhonebookBase::iterator> range =
        destPhonebook->equal_range(text);
      
      SortedPhonebookBase::iterator first = range.first;
      if (first != destPhonebook->end() && range.second == ++first)
      {                         // just one text in the destPhonebook
        if (! (*range.first == *i)) // overwrite if different in destination
        {
          if (verbose)
            cout << stringPrintf(_("updating '%s' tel# %s to new tel# %s"),
                                 range.first->text().c_str(),
                                 range.first->telephone().c_str(),
                                 i->telephone().c_str())
                 << endl;
          
          *range.first = *i;
        }
      }
      lastText = text;
    }
    firstLoop = false;
  }
}

// the same but for indexed phonebooks

void updateEntriesIndexed(SortedPhonebookRef sourcePhonebook,
                          SortedPhonebookRef destPhonebook,
                          bool verbose)
{
  for (SortedPhonebookBase::iterator i = sourcePhonebook->begin();
       i != sourcePhonebook->end(); ++i)
  {
    int index = i->index();
    
    SortedPhonebookBase::iterator j = destPhonebook->find(index);
    
    if (j != destPhonebook->end())
    {                           // index present in the destPhonebook
      if (! (*j == *i))         // overwrite if different in destination
      {
        if (verbose)
          cout << stringPrintf(_("updating '%s' tel# %s to new tel# %s"
                                 "(index %d)"),
                               j->text().c_str(),
                               j->telephone().c_str(),
                               i->telephone().c_str(), i->index())
               << endl;
        
        *j = *i;
      }
    }
  }
}

// delete those entries from destPhonebook, that are not present
// in sourcePhonebook

void deleteNotPresent(SortedPhonebookRef sourcePhonebook,
                      SortedPhonebookRef destPhonebook,
                      bool indexed, bool verbose)
{
  for (SortedPhonebookBase::iterator i = destPhonebook->begin();
       i != destPhonebook->end(); ++i)
  {
    pair<SortedPhonebookBase::iterator, SortedPhonebookBase::iterator> range;
    if (indexed)
    {
      int index = i->index();
      range = sourcePhonebook->equal_range(index);
    }
    else
    {
      string text = i->text();
      range = sourcePhonebook->equal_range(text);
    }
        
    bool found = false;
    for (SortedPhonebookBase::iterator j = range.first;
         j != range.second; ++j)
    {
      i->setUseIndex(indexed);
      if (j->telephone() == i->telephone())
      {
        found = true;
        break;
      }
    }
    if (! found)
    {
      if (verbose)
      {
        cout << stringPrintf(_("deleting '%s' tel# %s"),
                             i->text().c_str(), i->telephone().c_str());
        if (indexed)
          cout << stringPrintf(_(" (index #%d)"), i->index());
        cout << endl;
      }
      destPhonebook->erase(i);
#ifdef BUGGY_MAP_ERASE
	  deleteNotPresent(sourcePhonebook, destPhonebook, indexed, verbose);
	  return;
#endif
    }
  }
}

// *** main program

int main(int argc, char *argv[])
{
  try
  {
    // handle command line options
    string destination;
    string source;
    string destinationBackend;
    string sourceBackend;
    string baudrate;
    bool doSynchronize = true;
    string phonebook;
    SortedPhonebookRef sourcePhonebook, destPhonebook;
    bool verbose = false;
    bool indexed = false;
    string initString = DEFAULT_INIT_STRING;
    bool swHandshake = false;
    string charSet;
    Ref<MeTa> sourceMeTa, destMeTa;

    int opt;
    int dummy;
    while((opt = getopt_long(argc, argv, "I:p:s:d:b:cyhvViD:S:Xt:", longOpts,
                             &dummy))
          != -1)
      switch (opt)
      {
      case 'X':
        swHandshake = true;
        break;
      case 'I':
        initString = optarg;
        break;
      case 'V':
        verbose = true;
        break;
      case 'p':
        phonebook = optarg;
        break;
      case 'd':
        destination = optarg;
        break;
      case 's':
        source = optarg;
        break;
      case 'D':
        destinationBackend = optarg;
        break;
      case 'S':
        sourceBackend = optarg;
        break;
      case 'b':
        baudrate = optarg;
        break;
      case 't':
        charSet = optarg;
        break;
      case 'c':
        doSynchronize = false;
        break;
      case 'i':
        indexed = true;
        break;
      case 'y':
        doSynchronize = true;
        break;
      case 'v':
        cerr << argv[0] << stringPrintf(_(": version %s [compiled %s]"),
                                        VERSION, __DATE__) << endl;
        exit(0);
        break;
      case 'h':
        cerr << argv[0] << _(": [-b baudrate][-c][-d device or file][-h]"
                             "[-I init string]\n"
                             "  [-p phonebook name][-s device or file]"
                             "[-t charset][-v]"
                             "[-V][-y][-X]") << endl
             << endl
             << _("  -b, --baudrate    baudrate to use for device "
                  "(default: 38400)")
             << endl
             << _("  -c, --copy        copy source entries to destination")
             << endl
             << _("  -d, --destination sets the destination device to "
                  "connect \n"
                  "                    to, or the file to write") << endl
             << _("  -D, --destination-backend sets the destination backend")
             << endl
             << _("  -h, --help        prints this message") << endl
             << _("  -i, --index       takes index positions into account")
             << endl
             << _("  -I, --init        device AT init sequence") << endl
             << _("  -p, --phonebook   name of phonebook to use") << endl
             << _("  -s, --source      sets the source device to connect to,\n"
                  "                    or the file to read") << endl
             << _("  -t, --charset     sets the character set to use for\n"
                  "                    phonebook entries") << endl
             << _("  -S, --source-backend sets the source backend")
             << endl
             << _("  -v, --version     prints version and exits") << endl
             << _("  -V, --verbose     print detailed progress messages")
             << endl
             << _("  -X, --xonxoff     switch on software handshake") << endl
             << _("  -y, --synchronize synchronize destination with source\n"
                  "                    entries (destination is overwritten)\n"
                  "                    (see gsmpb(1) for details)")
             << endl << endl;
        exit(0);
        break;
      case '?':
        throw GsmException(_("unknown option"), ParameterError);
        break;
      }

    // check if all parameters all present
    if (destination == "" || source == "")
      throw GsmException(_("both source and destination must be given"),
                         ParameterError);

    // start accessing source mobile phone or file
    if (sourceBackend != "")
      sourcePhonebook =
        CustomPhonebookRegistry::createPhonebook(sourceBackend, source);
    else if (source == "-")
      sourcePhonebook = new SortedPhonebook(true, indexed);
    else if (isFile(source))
      sourcePhonebook = new SortedPhonebook(source, indexed);
    else
    {
      if (phonebook == "")
        throw GsmException(_("phonebook name must be given"), ParameterError);

      sourceMeTa = new MeTa(new
#ifdef WIN32
                            Win32SerialPort
#else
                            UnixSerialPort
#endif
                            (source,
                             baudrate == "" ? DEFAULT_BAUD_RATE :
                             baudRateStrToSpeed(baudrate), initString,
                             swHandshake));
      if (charSet != "")
        sourceMeTa->setCharSet(charSet);
      sourcePhonebook =
        new SortedPhonebook(sourceMeTa->getPhonebook(phonebook));
    }

    // make sure destination.c_str file exists
	if (destination != "")
	{
      try
	  {
        ofstream f(destination.c_str(), ios::out | ios::app);
	  }
      catch (exception)
	  {
	  }
	}

    // start accessing destination mobile phone or file
    if (destinationBackend != "")
      destPhonebook =
        CustomPhonebookRegistry::createPhonebook(destinationBackend,
                                                 destination);
    else if (destination == "-")
      destPhonebook = new SortedPhonebook(false, indexed);
    else if (isFile(destination))
      destPhonebook = new SortedPhonebook(destination, indexed);
    else
    {
      if (phonebook == "")
        throw GsmException(_("phonebook name must be given"), ParameterError);

      destMeTa = new MeTa(new 
#ifdef WIN32
                          Win32SerialPort
#else
                          UnixSerialPort
#endif
                          (destination,
                           baudrate == "" ? DEFAULT_BAUD_RATE :
                           baudRateStrToSpeed(baudrate), initString,
                           swHandshake));
      if (charSet != "")
        destMeTa->setCharSet(charSet);
      PhonebookRef destPb = destMeTa->getPhonebook(phonebook);

      // check maximum lengths of source text and phonenumber when writing to
      // mobile phone
      unsigned int maxTextLen = destPb->getMaxTextLen();
      unsigned int maxTelLen = destPb->getMaxTelephoneLen();

      for (SortedPhonebookBase::iterator i = sourcePhonebook->begin();
           i != sourcePhonebook->end(); ++i)
        if (i->text().length() > maxTextLen)
          throw GsmException(
            stringPrintf(_("text '%s' is too large to fit into destination "
                           "(maximum size %d characters)"),
                         i->text().c_str(), maxTextLen),
            ParameterError);
        else if (i->telephone().length() > maxTelLen)
          throw GsmException(
            stringPrintf(_("phone number '%s' is too large to fit into "
                           "destination (maximum size %d characters)"),
                         i->telephone().c_str(), maxTelLen),
            ParameterError);

      // read phonebook
      destPhonebook = new SortedPhonebook(destPb);      
    }

    // now do the actual work
    if (doSynchronize)
    {                           // synchronizing
      if (indexed)
      {
        sourcePhonebook->setSortOrder(ByIndex);
        destPhonebook->setSortOrder(ByIndex);
        // for an explanation see below
        updateEntriesIndexed(sourcePhonebook, destPhonebook, verbose);
        deleteNotPresent(sourcePhonebook, destPhonebook, true, verbose);
        insertNotPresent(sourcePhonebook, destPhonebook, true, verbose);
      }
      else
      {
        sourcePhonebook->setSortOrder(ByText);
        destPhonebook->setSortOrder(ByText);
        // the following is done to avoid superfluous writes to the TA
        // (that takes time) and keep updated (ie. telephone number changed)
        // entries at the same place
        // 1. update entries in place where just the number changed
        updateEntries(sourcePhonebook, destPhonebook, verbose);
        // 2. delete those that are not present anymore
        deleteNotPresent(sourcePhonebook, destPhonebook, false, verbose);
        // 3. insert the new ones
        insertNotPresent(sourcePhonebook, destPhonebook, false, verbose);
      }
    }
    else
    {                           // copying
      destPhonebook->clear();
      for (SortedPhonebookBase::iterator i = sourcePhonebook->begin();
           i != sourcePhonebook->end(); ++i)
      {
        if (verbose)
        {
          cout << stringPrintf(_("inserting '%s' tel# %s"),
                               i->text().c_str(), i->telephone().c_str());
          if (indexed)
            cout << stringPrintf(_(" (index #%d)"), i->index());
          cout << endl;
        }
        destPhonebook->insert(*i);
      }
    }
  }
  catch (GsmException &ge)
  {
    cerr << argv[0] << _("[ERROR]: ") << ge.what() << endl;
    return 1;
  }
  return 0;
}
